### DISCUSSION TITLE
Best practices for passing url list

### DISCUSSION THREAD

**Discussion by benjohnsonn** [harlan-zw/unlighthouse#178] [Q&A]
Hey y'all!

I've integrated unlighthouse with a Scrapy crawler I currently use – instead of using unlighthouse's crawling function, I'm feeding it a list of URLs from Scrapy, ranging from 200 to 10,000 URLs. My approach follows the advice in the docs section about [Manually Providing URLs](https://unlighthouse.dev/guide/guides/url-discovery#manually-providing-urls). 

My first question is whether i'm still leveraging unlighthouse's efficiency when I bypass the crawling feature?

Since I'm passing exact URLs, I'm not sure if route sampling would still apply. Would it be better to convert my URL list into relative path folders with regex rules and then still use unlighthouse's crawler? 

The main goal is to replace an existing python script I wrote to directly query the lighthouse api, which works, but isn't performant at all.

I know this isn't really the intended use case, so any pointers or insights you can provide would be greatly appreciated!

Additionally, My second question is, if it's possible to configure the csvExpanded to include even more information per url? It seems that the UI crawl results capture more information.

**Comment by Charlie284** [comment]
You are still getting Unlighthouse's queueing, isolated browser workers, report UI, and CI reporters. Supplying the URLs yourself only removes the discovery work. It does not make each Lighthouse audit cheaper.

There are two important current behaviors to account for:

1. An explicit `urls` list [disables the sitemap, robots lookup, crawler, and dynamic sampling](https://github.com/harlan-zw/unlighthouse/blob/d3a23ad98c0bc92f959b00e07d8ddab6e05eea90/packages/core/src/discovery/routes.ts#L24-L40). Therefore, no, route sampling does not apply to that list.
2. `scanner.maxRoutes` defaults to 200. Unless you raise it or set it to `false`, only 200 routes will be queued.

For thousands of paths, I would not put them in `--urls`; a 10,000-item command line is likely to hit command-length limits. Load Scrapy's output as an array in the config instead:

```ts
// unlighthouse.config.ts
import { readFileSync } from 'node:fs'
import { defineUnlighthouseConfig } from 'unlighthouse/config'

const urls = JSON.parse(readFileSync('./urls.json', 'utf8')) as string[]

export default defineUnlighthouseConfig({
  site: 'https://example.com',
  // urls.json contains an array of same-origin relative paths.
  urls,
  scanner: {
    maxRoutes: false,
    samples: 1,
  },
  puppeteerClusterOptions: {
    maxConcurrency: 2,
  },
})
```

The docs also show a zero-argument async `urls` function, but current v0.18.0 checks `resolvedConfig.urls?.length` before deciding whether to call it. A zero-argument function has `length === 0`, so the synchronous array form above avoids that current source/docs mismatch.

For 10,000 URLs I would still split `urls.json` into deterministic batches and give each `unlighthouse-ci` run a different `outputPath`. One process holding 10,000 route reports is a large failure domain, and 10,000 exact URLs still means 10,000 Lighthouse runs. Keep concurrency conservative if the performance numbers matter because competing Chrome instances can skew them.

Converting the list to regexes and turning the crawler back on would add discovery work and would not preserve your exact list. If representative coverage is acceptable, group/sample the Scrapy URLs by page template before passing them in. If every URL must be tested, keep the explicit list.

`csvExpanded` is driven by `client.columns`, so you can add another numeric or binary Lighthouse audit to the CSV. For example:

```ts
hooks: {
  'resolved-config'(config) {
    config.client.columns.performance.push({
      label: 'Server response time',
      key: 'report.audits.server-response-time',
      sortKey: 'numericValue',
      cols: 1,
    })
  },
},
```

Then run `unlighthouse-ci --reporter csvExpanded`. The limitation is that the current reporter [only emits score-like audit columns](https://github.com/harlan-zw/unlighthouse/blob/d3a23ad98c0bc92f959b00e07d8ddab6e05eea90/packages/cli/src/reporters/csvExpanded.ts#L6-L59); it filters out informative/not-applicable audits and does not expose arbitrary UI crawl fields such as the full SEO object. For complete Lighthouse data, consume each route's `lighthouse.json`. For Unlighthouse-specific HTML/SEO fields shown in the UI, a small programmatic exporter from `worker.reports()` is the appropriate route rather than trying to force them through `csvExpanded`.
